### PR TITLE
[JIT] add support for ModuleDict

### PR DESCRIPTION
**Pull request by eellison**
Add support for nn.ModuleDict in script. This is needed to support torchvision.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -7323,6 +7323,58 @@ def forward(self):
...
+
+                for mod in self.moduledict.values():
+                    x = mod(x)
+
+                for key in self.moduledict.keys():
+                    names.append(key)
+
+                return x, names
+
+        m = M()
```
Comment (driazati) [inline test/test_jit.py]: `checkModule` does this same thing

https://github.com/pytorch/pytorch/blob/7ab4ad7b6df5a52719e7d3c4012403e05a749f2c/test/test_jit.py#L14837

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 2]
```diff
@@ -276,6 +276,53 @@ Value* ModuleValue::asValue(const SourceRange& loc, Function& m) {
   return self_;
 }
 
+std::vector<std::shared_ptr<SugaredValue>> ModuleValue::desugarModuleContainer(
+    bool get_keys,
+    bool get_values,
+    const SourceRange& loc,
+    Function& m) {
+  // the submodules in the module list may be a mix of python objects
```
Comment (driazati) [inline torch/csrc/jit/script/python_sugared_value.cpp]: Don't all the elements of `ModuleList/Dict/Sequential` get converted to `ScriptModule`s, so there aren't Python objects in any module containers?

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 3]
```diff
@@ -316,6 +363,26 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
...
+      bool get_values = false;
+      if (field == "items") {
+        get_keys = true;
+        get_values = true;
+      } else if (field == "values") {
+        get_values = true;
+      } else {
+        get_keys = true;
+      }
+      return std::make_shared<ConstantTupleMethod>(
```
Comment (driazati) [inline torch/csrc/jit/script/python_sugared_value.cpp]: Why is this a `ConstantTupleMethod` if `ModuleDict`s can't be called?

**Review by driazati — COMMENTED** [review]
Mostly good, a couple small questions

File: torch/csrc/jit/script/python_sugared_value.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/script/python_sugared_value.cpp]: Not in the case of `ScriptModules`. This is a refactoring of the existing method.

File: torch/csrc/jit/script/python_sugared_value.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/script/python_sugared_value.cpp]: What do you mean? It's not being called, a function on it is being called.

File: test/test_jit.py [hunk 4]
```diff
@@ -7323,6 +7323,58 @@ def forward(self):
...
+                for name in self.moduledict:
+                    names.append(name)
+
+                for name, mod in self.moduledict.items():
+                    if name != skip_name:
+                        names.append(name)
+                        x = mod(x)
+
+                for mod in self.moduledict.values():
+                    x = mod(x)
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/test_jit.py]: Well, the three modules that i have aren't commutative so it's checking that the order is correct.

**Review by driazati — APPROVED** [review]

**Pull request closed without merging**